### PR TITLE
Add Zstandard

### PR DESCRIPTION
**Pull request by apfel**
Added Zstandard, it should be a good alternative to zlib.
I've added a sample and tested it on my Xbox and didn't encounter any issues.

Edit: maybe a sample showcasing how to add a library in a subdirectory within a project would be useful in case someone makes an application that requires another not-yet-ported library.

### CODE REVIEW

**Comment by dracc** [comment]
nit: Commit message doesn't follow format `section: Description`. Suggestion for new message: `zstd: Add zstd <version number>`.

File: lib/zstd/Makefile
Comment (thrimbor) [inline lib/zstd/Makefile:9]: `_byteswap_ulong` and `_byteswap_uint64` are regular functions on Windows, so I'd recommend putting their declarations in `stdlib_ext_.h` and implement them in `stdlib_ext_.c` to avoid introducing additional compiler parameters. You can still use the builtins for the implementation, of course.

**Review by thrimbor — CHANGES_REQUESTED** [review]
Is there a specific use-case you have in mind for this? zlib was only introduced as a dependency of libpng, which in turn is a dependency of SDL_image.

**Comment by Apfel** [comment]
I needed it for a project because I find its interface a little easier to use (also I had issues compiling zlib on Windows), and considering zstd was easy to port and would maybe be useful for others too, I figured adding this would be a good idea.

It would be cool if the nxdk had a portlibs or other section for ready-to-use ported libraries, I guess.

File: samples/compression-zstd/Makefile
Comment (thrimbor) [reply samples/compression-zstd/Makefile:3]: Wildcard is still there.

File: lib/zstd/Makefile
Comment (thrimbor) [reply lib/zstd/Makefile]: You moved the definitions into the header, but I meant that they should be actual functions, like they are on Windows.

**Comment by thrimbor** [comment]
Please don't mark maintainer comments as resolved yourself, it makes it harder to keep track of things.

> It would be cool if the nxdk had a portlibs or other section for ready-to-use ported libraries, I guess.

Yeah, I've been thinking about the best way to handle ported libraries - we can't keep adding them like this forever, but so far I haven't found the ideal solution yet.

**Comment by Apfel** [comment]
> Please don't mark maintainer comments as resolved yourself, it makes it harder to keep track of things.

Sorry about that.

> Yeah, I've been thinking about the best way to handle ported libraries - we can't keep adding them like this forever, but so far I haven't found the ideal solution yet.

What about an `nxdk-portlibs` repository? We put all the ported libraries there and thus keep them separate from the nxdk's core.

**Comment by thrimbor** [comment]
> What about an `nxdk-portlibs` repository? We put all the ported libraries there and thus keep them separate from the nxdk's core. I'd suggest moving everything third-party and/or optional there, including things like SDL.

That's one of the things I had in mind, but there are limitations caused by the way we currently build things that probably need to get resolved first.

Are you still working on this? If yes, can you integrate your fixes into the commits where the code was added? Some fixes that are required: Your last commit title is too long to be displayed properly, the sample is using a wildcard (`$(shell find . -name '*.c')`) to find the code, and `_byteswap_ushort` etc. should become regular functions.

**Comment by Apfel** [comment]
*afaik* `_byteswap_ulong` and `__builtin_bswap32` (as well as the others) have the same behavior, it's one of these Microsoft-really-wants-to-be-different cases.

**Comment by Apfel** [comment]
> can you integrate your fixes into the commits where the code was added?
If you tell me how, I can try to do that.

**Comment by JayFoxRox** [comment]
@thrimbor user is waiting for a response regarding git history.

File: lib/xboxrt/libc_extensions/stdlib_ext_.h
Comment (JayFoxRox) [inline lib/xboxrt/libc_extensions/stdlib_ext_.h:37]: #433 was merged recently, so the win32 byteswap functions no longer have to be a part of this.

File: .gitmodules
Comment (Apfel) [reply .gitmodules:36]: ...is that really important?

**Comment by JayFoxRox** [comment]
I just tried to "port" zstd with https://github.com/JayFoxRox/nxdk/pull/74 and (after a tiny change), this is all that's required to "port" it:
```
git clone git clone https://github.com/facebook/zstd.git
cd zstd/build/cmake
mkdir build
nxdk-cmake .. -DZSTD_BUILD_PROGRAMS=OFF -DCMAKE_INSTALL_PREFIX=$NXDK_DIR/usr/
make
make install
```
*(The install prefix is subject to change + it might not even be necessary in future revisions; I also didn't test it)*

Because there's no nxdk-specific stuff, this is *much* more maintainable and future proof.

I feel like we should close this PR (unmerged) and focus on making nxdk more usable with standard tools to avoid the need for this kind of PRs. We already just had issues to upgrade / maintain a dependency in #456.

So while I appreciate the work that has gone into this PR (thanks @Apfel!), this won't be a scalable solution for us (not this library in particular, but the approach of hosting ports on this repository).

Rather than adding new ports, I'd prefer to get rid of them (before people depend on it with our weird build-system).

File: .gitmodules
Comment (JayFoxRox) [reply .gitmodules:36]: It had been added recently, so it's only missing in this outdated branch. However, don't bother with it, my opinion (and proposal) is in https://github.com/XboxDev/nxdk/pull/418#issuecomment-822691345.

I hope you agree that my proposal is a better solution (and that it hopefully also suits your needs) than to continue with this PR.

*(If another maintainer wants to fixup, review and maintain this, that's fine with me, but I think it's unlikely)*

**Comment by Apfel** [comment]
We've already agreed upon making a `portlibs` sorta-ish repository for ported libraries instead (perhaps taking advantage of `pacman` would do it for us? devkitPro did this with their toolchains and it works quite well) to prevent the nxdk from getting clogged with tons of libraries. Therefore, if `nxdk-cmake` can take care of Zstandard and other CMake-based libraries with little to no additional effort required, I'll have no problem with dropping this; this PR mainly waited until we're in the clear to start making the aforementioned repository.

As things stand, we should definitely focus on improving the build system.

I'll leave closing this pull request to you.

**Comment by sharkwouter** [comment]
The PSPDEV SDK for the Playstation Portable also uses pacman. We decided to make a small patch for upstream pacman and use that with a wrapper. I made a repository which hosts a pacman repository as a release too, so users can update their libraries by running ``psp-pacman -Syu``or install new ones that way.

Here is the repository with the libraries: https://github.com/pspdev/psp-packages

This is the one with pacman: https://github.com/pspdev/psp-pacman

Might be interesting for reference.

**Comment by thrimbor** [comment]
Using pacman was my first thought as well, but I ultimately decided against it.

Given the Xbox's small amount of memory and the fact that everything is statically linked makes it quite desirable to rely on LTO as much as possible. Unfortunately llvm's bytecode files are version-specific, so we cannot ship libraries prebuilt with LTO (unless we control the compiler version like in the docker image).
We'd have to use a package manager that can build from source if necessary, use binary packages if possible, and supports build options (like enabling LTO, or enabling certain library features) that influence this selection.

I didn't find a solution I liked, so I prototyped my own package manager last year. It's supposed to be a bit of a mixture of Gentoo's Portage and Arch's Pacman, but with a few other ideas of mine, like having safely parsable build descriptions files. It's not ready for regular usage though, it's a bunch of hacky Python code, and I'm not even sure I want to use Python for the final product.

unknown2
unknown3
unknown4
unknown5

**Comment by JayFoxRox** [comment]
Please move the package-manager discussion elsewhere.

**Comment by thrimbor** [comment]
The PR author has agreed to closing this on Discord. The library will eventually be provided as a package once we have a package manager.

**Pull request closed without merging**
